### ISSUE TITLE
predefining SERVER_NAME doesn't allow for flask serving multiple domains

### ISSUE THREAD

**Issue by Strahlemann** [h-janes/flask-sitemapper#2]
Hi there,
When defining SERVER_NAME in the app config, a limitation is introduced if your same flask app serves multiple domains.
Would it be possible to use the request.host element in the generator instead to generate the URLs when the sitemap is called?

I tried to get it done myself, but I have to admit I'm getting lost here.

**Comment by h-janes** [comment]
Hi, good idea, I will look into this

**Comment by h-janes** [comment]
It seems this limitation is removed by using `with self.app.test_request_context():` instead of `with self.app.app_context():` in line 65 of sitemapper.py

I will do some tests and create a new release with the changes. Thank you!

**Comment by Strahlemann** [comment]
Hi there,
I played around a bit more, and I think it is better to create the URIs when running the generator function.
That way, you always get the URI of the initialized flask instance which supports multiple domains based on the flask.request.

Code proposal:
```python
from typing import Callable, Union
from functools import wraps
from jinja2 import Environment, BaseLoader
from flask import Flask, url_for, Response
from .templates import SITEMAP, SITEMAP_INDEX


class Sitemapper:
    def __init__(self, app: Flask = None, https: bool = True, master: bool = False) -> None:
        self.app = None
        self.urlset = []
        self.endpoints = []
        self.scheme = "https" if https else "http"
        self.template = SITEMAP_INDEX if master else SITEMAP
        self.deferred_functions = []

        if app:
            self.init_app(app)

    def init_app(self, app: Flask) -> None:
        """A conventional interface allowing initializing a flask app later"""
        self.app = app

        for deferred in self.deferred_functions:
            deferred(self)

        self.deferred_functions.clear()

    def include(self, **kwargs) -> Callable:
        """A decorator for view functions that calls `add_endpoint`"""

        def decorator(func: Callable) -> Callable:
            self.add_endpoint(func, **kwargs)

            @wraps(func)
            def wrapper(*args, **kwargs):
                with self.app.app_context():
                    return func(*args, **kwargs)

            return wrapper

        return decorator

    def find_endpoint_from_view_func(self, func: Callable) -> str:
        """Find the endpoint name from a list of registered view functions"""
        for endpoint, view_func in self.app.view_functions.items():
            if func is view_func:  # compare ids of function objects
                return endpoint

        # func not registered as view func
        raise ValueError(f"function {func.__name__} in module {func.__module__} is not a registered view function")

    def add_endpoint(self, view_func: Union[Callable, str], **kwargs) -> None:
        """Adds the URL of `view_func` to the sitemap with any provided arguments"""
        # if flask app is not yet initialized, then register a deferred function and run it later in init_app()
        if not self.app:
            self.deferred_functions.append(lambda s: s.add_endpoint(view_func, **kwargs))
            return

        if not isinstance(view_func, str):
            endpoint = self.find_endpoint_from_view_func(view_func)
        else:
            endpoint = view_func

        self.endpoints.append(endpoint)

    def generate(self) -> Response:
        """Creates a flask `Response` object for the sitemap's view function"""
        template = Environment(loader=BaseLoader).from_string(self.template)
        for endpoint in self.endpoints:
            url = {"loc": url_for(endpoint, _external=True, _scheme=self.scheme)}
            self.urlset.append(url)
        xml = template.render(urlset=self.urlset)
        return Response(xml, content_type="application/xml")
```

**Comment by h-janes** [comment]
Hi, thanks for the feedback and code proposal.

Have you found that the current version using `self.app.test_request_context()` to generate the URLs is still not properly supporting apps with multiple domains?

I will look into this and work on your suggestion over the next few days

**Comment by Strahlemann** [comment]
Sorry - I should have mentioned that.
Using your suggestion ended up in localhost as the base_url (sheme://localhost/endpoint).

This might be the combination of venv and gunicorn with nginx causing that so I was looking for a way to generate the url’s based on the current instance (pointing three domains to the same flask page).

**Comment by h-janes** [comment]
Thanks,

I have created [a branch](https://github.com/h-janes/flask-sitemapper/tree/2-predefining-server_name-doesnt-allow-for-flask-serving-multiple-domains) for this issue and have made changes based on your suggestion in that branch. I have used a slightly different approach than your proposal in order to keep the lastmod, changefreq, and priority arguments working and include recent commits.

The URLs are now stored as `URL` objects rather than dicts, with methods that generate their URL loc and XML during sitemap generation. The changes are in `sitemapper.py` and `templates.py`

Does the version on this branch work in your use case now?
